peer/peer.go

```diff
@@ -114,15 +153,37 @@ func parsePiece(payload []byte) (*chunk, error) {
 	// index of the message (4 byte big endian)
 	// beginning of the chunk (4 byte big endian)
 	// payload
-	if len(payload) < 0 {
+	if len(payload) < 8 {
 		return nil, fmt.Errorf("expected message of length at least 8 got %d instead", len(payload))
 	}
 	index := binary.BigEndian.Uint32(payload[:4])
 	begin := binary.BigEndian.Uint32(payload[4:8])
 	return &chunk{
-		Index: int(index),
-		Begin: int(begin),
-		Value: payload[8:],
+		chunkType: cFile,
+		index:     int(index),
+		begin:     int(begin),
+		value:     payload[8:],
+	}, nil
+}
+
+// parseExtended parse an extended message as a chunk
+func (p *peer) parseExtended(payload []byte) (*chunk, error) {
+	// an extended message has the following format:
+	// uint8 extended message id
+	// bencoded dictionary
+	// piece data (16 KiB unless it is the last piece)
+	if len(payload) < 1 {
+		return nil, fmt.Errorf("expected message of length at least 1 got %d instead", len(payload))
+	}
+	parsed, index, err := fileutils.ParseExtensionsMetadata(payload[1:])
+	if err != nil || parsed == nil {
+		return nil, err
+	}
+	return &chunk{
+		chunkType: cInfo,
+		index:     index,
+		begin:     index * chunkSize,
```

Comment:
All metainfo messages should be 16 KiB long (minus the last); create another constant to make it clearer?